engine/src/main/java/org/hibernate/validator/internal/engine/ConstraintViolationImpl.java

```diff
@@ -278,28 +278,28 @@ public boolean equals(Object o) {
 		if ( interpolatedMessage != null ? !interpolatedMessage.equals( that.interpolatedMessage ) : that.interpolatedMessage != null ) {
 			return false;
 		}
-		if ( propertyPath != null ? !propertyPath.equals( that.propertyPath ) : that.propertyPath != null ) {
+		if ( messageTemplate != null ? !messageTemplate.equals( that.messageTemplate ) : that.messageTemplate != null ) {
 			return false;
 		}
-		if ( rootBean != null ? !rootBean.equals( that.rootBean ) : that.rootBean != null ) {
+		if ( propertyPath != null ? !propertyPath.equals( that.propertyPath ) : that.propertyPath != null ) {
 			return false;
 		}
-		if ( leafBeanInstance != null ? !leafBeanInstance.equals( that.leafBeanInstance ) : that.leafBeanInstance != null ) {
+		if ( rootBean != null ? ( rootBean != that.rootBean ) : that.rootBean != null ) {
 			return false;
 		}
-		if ( constraintDescriptor != null ? !constraintDescriptor.equals( that.constraintDescriptor ) : that.constraintDescriptor != null ) {
+		if ( leafBeanInstance != null ? ( leafBeanInstance != that.leafBeanInstance ) : that.leafBeanInstance != null ) {
 			return false;
 		}
-		if ( elementType != null ? !elementType.equals( that.elementType ) : that.elementType != null ) {
+		if ( value != null ? ( value != that.value ) : that.value != null ) {
 			return false;
 		}
-		if ( messageTemplate != null ? !messageTemplate.equals( that.messageTemplate ) : that.messageTemplate != null ) {
+		if ( constraintDescriptor != null ? !constraintDescriptor.equals( that.constraintDescriptor ) : that.constraintDescriptor != null ) {
 			return false;
 		}
-		if ( rootBeanClass != null ? !rootBeanClass.equals( that.rootBeanClass ) : that.rootBeanClass != null ) {
+		if ( elementType != null ? !elementType.equals( that.elementType ) : that.elementType != null ) {
 			return false;
 		}
-		if ( value != null ? !value.equals( that.value ) : that.value != null ) {
+		if ( rootBeanClass != null ? !rootBeanClass.equals( that.rootBeanClass ) : that.rootBeanClass != null ) {
```

Comment:
One more thing .... Do we need to check this at all ? if rootBean != that.rootBean  CV are not equals and if rootBean == that.rootBean classes should already be equal so no need to check, right ?

Reply:
Hmmm, yes, I think you're right.